### ISSUE TITLE
Find ranges

### ISSUE THREAD

**Issue by wertzui** [morelinq/MoreLINQ#631]
It would be handy to have a method to find ranges of consecutive Numbers (int or long) in a given (ordered) collection.

Something like `public static IEnumerable<Tuple<long, long>> FindConsecutiveRanges(this IEnumerable<long> source)`

**Comment by atifaziz** [comment]
I believe you can achieve this with a combination of [`Lag`](https://morelinq.github.io/3.2/ref/api/html/Overload_MoreLinq_MoreEnumerable_Lag.htm) and [`Segment`](https://morelinq.github.io/3.2/ref/api/html/Overload_MoreLinq_MoreEnumerable_Segment.htm):

```c#
var nums = new[] { -5, 1, 2, 3, 15, 16, 17, 95, 100, 101, 102, 103, 104 };

var ranges =
    nums.Select(n => (Some: true, Num: n))
        .Lag(1, (curr, prev) => (Curr: curr, Prev: prev))
        .Segment(e => !e.Prev.Some || e.Curr.Num - e.Prev.Num != 1)
        .Select(g => (First: g.First().Curr.Num, Last: g.Last().Curr.Num));
        
Console.WriteLine(string.Join(", ", ranges));
```

The output will be:

    (-5, -5), (1, 3), (15, 17), (95, 95), (100, 104)

One could then extrapolate the following generic implementation:

```c#
public static IEnumerable<(T, T)> ConsecutiveRanges<T>(this IEnumerable<T> source,
                                                       Func<T, T, bool> f) =>
    source.Select(n => (Some: true, Num: n))
          .Lag(1, (curr, prev) => (Curr: curr, Prev: prev))
          .Segment(e => !e.Prev.Some || !f(e.Prev.Num, e.Curr.Num))
          .Select(g => (First: g.First().Curr.Num, Last: g.Last().Curr.Num));
```

**Comment by Orace** [comment]
I think you can avoid the introduction of `Tuple` with `Lead` and `SkipLast`:

var nums = new[] { -5, 1, 2, 3, 15, 16, 17, 95, 100, 101, 102, 103, 104 };

    var ranges =
        nums.Lead(1, (curr, next) => (Curr: curr, Next: next))  // ValueTuple creation overload welcome here
            .SkipLast(1)
            .Segment(e => e.Next - e.Curr != 1)
            .Select(g => (First: g.First().Curr, Last: g.Last().Curr));
        
    Console.WriteLine(string.Join(", ", ranges));

Eye compiled, I can't test it right now.

**Comment by wertzui** [comment]
Both of your Solutions would enumerate the source collection multiple times.
But finding ranges can be achieved with just one enumeration.

**Comment by Orace** [comment]
> Both of your Solutions would enumerate the source collection multiple times.

I don't think so, both `Lead` and `Lag` use a `Queue` (here of size 1) in their implementation to avoid multiple iterations.

**Comment by wertzui** [comment]
> > Both of your Solutions would enumerate the source collection multiple times.
> 
> I don't think so, both `Lead` and `Lag` use a `Queue` (here of size 1) in their implementation to avoid multiple iterations.

One iteration to generate the range-collection and then each range-collection is iterated to generate the Tuples (technically First() will also generate an Enumerator but will immediately return the first element, the full iteration is done by Last()).

**Comment by Orace** [comment]
`Last()` is optimised for collection and O(1).
But as I said, the memory is indeed allocated and indeed the allocation is O(n).

**Comment by wertzui** [comment]
> `Last()` is optimised for collection and O(1).
> But as I said, the memory is indeed allocated and indeed the allocation is O(n).

I just had a look at the source of `Segment()` and it is using `List<T>` internally. So as long, as this internal implementation won't change, you are correct. But when for whatever reason this changes and is no longer using an indexable collection, it would become O(n).

**Comment by atifaziz** [comment]
> But we can use the other overload:
> 
> ```cs
> var nums = new[] { -5, 1, 2, 3, 15, 16, 17, 95, 100, 101, 102, 103, 104 };
> var ranges =
>     nums.Segment((current, previous,i) => current-previous != 1)
>         .Select(g => (First: g.First(), Last: g.Last()));
> Console.WriteLine(string.Join(", ", ranges));
> ```

@Orace Good one! I had completely forgotten about the other overload of `Segment` 🤦‍♂and so got distracted by `Lag`. Your version is far simpler!

> Both of your Solutions would enumerate the source collection multiple times.

@wertzui As you discovered, `Segment` yields lists so there is no re-iteration when fetching the first and last elements of a segment. #317 would have been helpful to assert that quickly.

> So as long, as this internal implementation won't change, you are correct. But when for whatever reason this changes and is no longer using an indexable collection, it would become O(n).

It cannot change, not within the frame of _LINQ to Objects_ or an in-memory execution model. A different _implementation_ of `Segement` can change that however; for example, one that operates remotely. This is the reason `Segment` returns a sequence of sequences and not a sequence of lists. Have a look at [my comment](https://github.com/morelinq/MoreLINQ/issues/98#issuecomment-250782442) on #98 for more background on this. I think it's a noble goal to not contractually bind other possible implementations in case we ever bring `IQueryable` into MoreLINQ but it hasn't happened in 10 years and I don't think it's going to happen in the foreseeable future either so I am beginning to reconsider the signatures to be more reflective and bound to the in-memory execution model. In fact, if you look at [`Window`](https://morelinq.github.io/3.2/ref/api/html/M_MoreLinq_MoreEnumerable_Window__1.htm), we already have that. You then wouldn't have to go popping the hood to look at what the implementation is really doing.

So we are down to just down to `Segement` + `Select`:

```
public static IEnumerable<(T First, T Last)>
    ConsecutiveRanges<T>(this IEnumerable<T> source, Func<T, T, bool> f) =>
        from g in source.Segment((curr, prev, _) => !f(prev, curr))
        select (g.First(), g.Last());
```

The only downside I see here is that each segment will be entirely committed to memory just to throw away all but its head and tail elements. That could be unacceptable in some scenarios. For example, I can imagine going over data files in a directory that contain time series data and validate that there are no “holes” in the timeline; in other words that there's only a consecutive segment. If the data across all files amounts to gigabytes in size then an approach using `Segment` could exhaust memory whereas [a hand-rolled and efficient implementation](https://github.com/morelinq/MoreLINQ/issues/631#issuecomment-549020767) wouldn't.

If we do this, I would go one step further and actually generalise `Segment` instead and make it even more powerful by describing how to _aggregate each segment_. The approach I see taking here is the same I did with [`Aggregate` over observable comprehensions](https://morelinq.github.io/3.2/ref/api/html/Overload_MoreLinq_Experimental_ExperimentalEnumerable_Aggregate.htm) (see #326 for background). In other words, it would come down to something like this:

```c#
var ranges =
    nums.Segment((curr, prev, i) => curr - prev != 1,
                 s => s.FirstAsync(), // from Rx
                 s => s.LastAsync(),  // from Rx
                 (fst, lst) => (First: fst, Last: lst));
```

If you also wanted a count then it would be just as simple as doing:

```c#
var ranges =
    nums.Segment((curr, prev, i) => curr - prev != 1,
                 s => s.FirstAsync(),
                 s => s.LastAsync(),
                 s => s.Count(),
                 (fst, lst, cnt) => (First: fst, Last: lst, Count: cnt));
```
